model/modeldecoder/v2/decoder.go

```diff
@@ -0,0 +1,231 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+func init() {
+	metadataWithKeyPool.New = func() interface{} {
+		return &metadataWithKey{}
+	}
+	metadataNoKeyPool.New = func() interface{} {
+		return &metadataNoKey{}
+	}
+}
+
+var metadataWithKeyPool, metadataNoKeyPool sync.Pool
+
+func fetchMetadataWithKey() *metadataWithKey {
+	return metadataWithKeyPool.Get().(*metadataWithKey)
+}
+func releaseMetadataWithKey(m *metadataWithKey) {
+	m.Reset()
+	metadataWithKeyPool.Put(m)
+}
+
+func fetchMetadataNoKey() *metadataNoKey {
+	return metadataNoKeyPool.Get().(*metadataNoKey)
+}
+func releaseMetadataNoKey(m *metadataNoKey) {
+	m.Reset()
+	metadataNoKeyPool.Put(m)
+}
+
+// DecodeProfileMetadata uses the given decoder to create the input models,
+// then runs the defined validations on the input models
+// and finally maps the values fom the input model to the given *model.Metadata instance
+func DecodeProfileMetadata(d decoder.Decoder, out *model.Metadata) error {
```

Comment:
I think the name DecodeProfileMetadata is too specific, since there's about profiles in this code; that just happens to be the only caller at the moment. Suggested renamings:

 - DecodeProfileMetadata -> DecodeMetadata (this is about decoding the Metadata struct fields)
 - DecodeMetadata -> DecodeNestedMetadata (this is about decoding a Metadata object nested in some other struct)

DecodeNestedMetadata expects `{"metadata":<metadata>}`, and behaves exactly the same as calling DecodeMetadata on the nested "metadata" field.

I think that's where you were going in https://github.com/elastic/apm-server/pull/4076 earlier - did I talk you out of using those names?

Reply:
I talked myself out of it, wasn't happy with either name, but I'll revert back to it.